name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/QuirionPDFExtractor.java

```diff
@@ -0,0 +1,377 @@
+                            t.setDateTime(asDate(v.get("date")));
+                            t.setShares(asShares(v.get("shares")));
+                            t.setCurrencyCode(asCurrencyCode(context.get("currency")));
+                            t.setAmount(asAmount(v.get("amount")));
+                        })
+                        .section("tax", "currency").optional()
+                        .match("^Erträgnisabrechnung, Ref.: \\d+ (\\d+.\\d+.\\d{4}+) (?<date>\\d+.\\d+.\\d{4}+) (?<amount>[\\d.]+,\\d{2})(.*)$")
+                        .match("^(.*)$")
+                        .match("^((Dis.oN\\n)|())(.{12}, ST) ([\\d.,]+)(.*)$")
+                        .match("^(.*)(KEST: )(?<currency>\\w{3}) -(?<tax>[\\d.]+,\\d{2}).*")
+                        .assign((t, v) -> {//t.addUnit(new Unit(Unit.Type.TAX, asAmount(v.get("tax"))));
+                                            Money taxes = Money.of(asCurrencyCode("EUR"),
+  //TODO other currencys not working                         Money.of(asCurrencyCode(v.get("currency")), asAmount(v.get("tax"))))))
+                                                            asAmount(v.get("tax")));
+                                            //.subtract(t.getMonetaryAmount()).toMoney();
+                                          t.addUnit(new Unit(Unit.Type.TAX, taxes));})
+                        //.assign((t, v) -> t.addUnit(new Unit(Unit.Type.TAX, asAmount(v.get("tax")))))
+//                                        Money.of(asCurrencyCode(v.get("currency")), asAmount(v.get("tax"))))))
+
+                        .section("tax", "currency").optional()
+                        .match("^Erträgnisabrechnung, Ref.: \\d+ (\\d+.\\d+.\\d{4}+) (?<date>\\d+.\\d+.\\d{4}+) (?<amount>[\\d.]+,\\d{2})(.*)$")
+                        .match("^((Dis.oN\\n)|())(.{12}, ST) ([\\d.,]+)(.*)$")
+                        .match("^(.*)(SOLI: )(?<currency>\\w{3}) -(?<tax>[\\d.]+,\\d{2})(.*)$")
+                        .assign((t, v) -> {//t.addUnit(new Unit(Unit.Type.TAX, asAmount(v.get("tax"))));
+                                            Money taxes = Money.of(asCurrencyCode("EUR"),
+  //TODO other currencys not working                         Money.of(asCurrencyCode(v.get("currency")), asAmount(v.get("tax"))))))
+                                                            asAmount(v.get("tax")));
+                                            //.subtract(t.getMonetaryAmount()).toMoney();
+                                          t.addUnit(new Unit(Unit.Type.TAX, taxes));})
+                        .wrap(t -> new TransactionItem(t)));
+        
+
+//Wertpapier Kauf, Ref.: 85249245 08.05.2019 10.05.2019 -235,86 EUR
+//SPDR MSCI Wrld Small Cap U.ETF Registered Shares o.N.
+//IE00BCBJG560, ST 3,648
+        
+        Block sellBlock = new Block("(Wertpapier Verkauf, Ref.: \\d+) (\\d+.\\d+.\\d{4}+) (\\d+.\\d+.\\d{4}+) ([\\d.]+,\\d{2}) (\\w{3}).*");
```

Comment:
Maybe adding an "endsWith" string could help here:
`
Block sellBlock = new Block("(Wertpapier Verkauf, Ref.: \\d+) (\\d+.\\d+.\\d{4}+) (\\d+.\\d+.\\d{4}+) ([\\d.]+,\\d{2}) (\\w{3}).*", ".* Ref\\.: .*");`

Reply:
Thanks for the hint, but this is causing that the block is cutted right behind the _Wertpapier Verkauf_ and not all bookings were find in the test. The easiest approach for me would be to limit the following lines in numbers.
Another approach would be to match with positive lookahead. Do you have an idea how to match a new line? I already tried:
`(.*\\n.*)+(?=(Sammelgutschrift)|(other booking types))`

Reply:
The matching logic searches line by line, not the entire text at once. That's why it is a bit tricky.

I think, it would make sense to change `PDFParser.java` in line 151 to this:

`endLine = findBlockEnd(lines, startLine + 1, endLine);`

because searching for the endsWith should start in the next line instead of the same as startsWith.

Reply:
I like this proposal. I would not allow for one-line blocks anymore, but I guess that is also fine. One line blocks are still possible if you match the next block on the next line.

Reply:
I think I have to correct myself. The way "endsWith" is implemented today, it only looks within the **current** block and if it does not find the **endsWith** it actually skips the transaction. So if a new block with ```".* Ref\\.: .*"``` is actually defined, then nothing is found. Needs more thinking....

Reply:
I have no implemented a "setMaxSize" method which limits a block to a maximum of this number of lines.

Reply:
> I have no implemented a "setMaxSize" method which limits a block to a maximum of this number of lines.

Thank you for that idea/feature, Andreas. That will be helpful in many other cases as well.